README.md
# node2nix-hs

[![Hackage](https://img.shields.io/hackage/v/node2nix-hs.svg?logo=haskell)](https://hackage.haskell.org/package/node2nix-hs)
[![MIT license](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
[![built with garnix](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fgarnix.io%2Fapi%2Fbadges%2Fptitfred%2Fnode2nix-hs)](https://garnix.io)

Partial drop&replace reimplementation of node2nix

Supports part of the original node2nix CLI options with notable differences:
- default value for `--composition` is now `node-dependencies.nix` and not `default.nix`
- short options  `-16` and `-18` are not supported
- node version up to 14 are not supported
- node version 20 is supported
- default node version is now 18

Following options are not implemented:
- option `--input`
- option `--development`
- option `--supplement-input`
- option `--supplement-output`
- flag `--include-peer-dependencies`
- flag `--no-flatten`
- option `--pkg-name`
- option `--registry`
- option `--registry-scope`
- option `--registry-auth-token`
- flag `--no-bypass-cache`
- flag `--use-fetchgit-private`
- flag `--strip-optional-dependencies`


CHANGELOG.md
# Changelog

`node2nix-hs` uses [PVP Versioning][1].

## 0.0.0.0

* Initially created.

[1]: https://pvp.haskell.org


test/Spec.hs
module Main
  ( main
  ) where

main :: IO ()
main = putStrLn "Tests for node2nix-hs"


src/CLI.hs
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE NamedFieldPuns    #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes       #-}
{-# LANGUAGE RecordWildCards   #-}

module CLI
  ( processLockFile
  , readOptions
  ) where

import           CLI.Options
import           Nix         hiding (Options)
import           Nix.TH      (nix)
import           Node2nixHs  (fromPackageLock)
import           NodeEnv     (nodeEnv)
import           PackageLock (PackageLock, loadPackageLock)

processLockFile :: Options -> IO ()
processLockFile options = do
  let Paths {.. } = paths options
  saveResult   packagesFile  =<< readPackageLock packageLockFile
  writeNixExpr compositionFile (composition options)
  unless (noCopyNodeEnv options) (writeNixExpr nodeEnvFile nodeEnv)

readPackageLock :: FilePath -> IO PackageLock
readPackageLock =
  let handleError s =
        die $ "Error while parsing the package-lock.json file: '" <> s <> "'"
   in either handleError pure <=< loadPackageLock

saveResult :: FilePath -> PackageLock -> IO ()
saveResult file = writeNixExpr file. fromPackageLock

writeNixExpr :: FilePath -> NExpr -> IO ()
writeNixExpr file expr =
  let prefix = "# This file has been generated by node2nix-hs 0.0.1. Do not edit!\n\n"
   in writeFile file (prefix ++ show (prettyNix expr))

toNixPackage :: NodeJSVersion -> Text
toNixPackage = \case
  NodeJS16 -> "nodejs-16_x"
  NodeJS18 -> "nodejs-18_x"
  NodeJS20 -> "nodejs-20_x"

composition :: Options -> NExpr
composition Options {..} =
  let nixpkgs = [nix| import <nixpkgs> { inherit system; } |]
      defaultNodePackage = "pkgs" @. toNixPackage nodeJSVersion
      params =
        mkParamSet
          [ ("pkgs"   , Just nixpkgs)
          , ("system" , Just [nix| builtins.currentSystem |])
          , ("nodejs" , Just defaultNodePackage)
          ]

      -- FIXME use callPackage instead
      nodeEnvArgs =
        [nix|
          {
            inherit (pkgs) stdenv lib python2 runCommand writeTextFile writeShellScript;
            inherit pkgs nodejs;
            libtool = if pkgs.stdenv.isDarwin then pkgs.darwin.cctools else null;
          }
        |]

      -- FIXME use callPackage instead
      nodePackagesArgs =
        [nix|
          {
            inherit (pkgs) fetchurl nix-gitignore stdenv lib fetchgit;
            inherit nodeEnv;
          }
        |]
      body =
        mkLets [ "nodeEnv" $= "import" @@ mkPath False (nodeEnvFile paths) @@ nodeEnvArgs ]
          $ "import" @@ mkPath False (packagesFile paths) @@ nodePackagesArgs
   in mkFunction params body


src/Node2nixHs.hs
{-# LANGUAGE DerivingStrategies         #-}
{-# LANGUAGE DuplicateRecordFields      #-}
{-# LANGUAGE FlexibleContexts           #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE LambdaCase                 #-}
{-# LANGUAGE NamedFieldPuns             #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE RecordWildCards            #-}

module Node2nixHs
  ( nodePackages
  , fromPackageLock
  , PackageDefinition(..)
  , PackageSource(..)
  , Dependency(..)
  , Source(..)
  , Hash(..)
  ) where

import qualified Data.Map.Strict as M
import           Data.Text       (replace)
import           Nix
import           PackageLock     (Package (..), PackageLock (..),
                                  packageNameFromPackageKey,
                                  selectRecursiveDependencies,
                                  selectRootPackages)
import           Text.Parsec     (anyChar, choice, many1, runParser, string)

data PackageDefinition =
  PackageDefinition
    { packageName  :: Text
    , src          :: FilePath
    , dependencies :: [Dependency]
    }

data Dependency = SimpleDependency    QName
                | RecursiveDependency QName [Dependency]

instance IsString Dependency where
  fromString = SimpleDependency . fromString

newtype QName = QName Text deriving newtype (IsString)

mkQName :: PackageSource -> QName
mkQName PackageSource { packageName, version } =
  QName (packageName <> "-" <> version)

data PackageSource =
  PackageSource
    { packageName :: Text
    , version     :: Text
    , source      :: Source
    }

data Source =
  Source
    { url  :: Text
    , hash :: Hash
    }

data Hash
  = SHA1   Text
  | SHA512 Text

hashFromIntegrity :: Text -> Maybe Hash
hashFromIntegrity =
  let runParsec p = runParser p () ""
      parser =
        choice
          [ string "sha512-" *> (SHA512 . fromString <$> many1 anyChar)
          , string "sha1-"   *> (SHA1   . fromString <$> many1 anyChar)
          ]
   in either (const Nothing) pure . runParsec parser

mkSources :: [PackageSource] -> NExpr
mkSources =
  let mkSingleSourceBinding ps = renderQName (mkQName ps) $= mkPackageSource ps
   in mkNonRecSet . fmap mkSingleSourceBinding

quote :: Text -> Text
quote s = "\"" <> s <> "\""

mkPackageSource :: PackageSource -> NExpr
mkPackageSource PackageSource { .. } =
  mkNonRecSet
    [ "name"        $= mkStr (nixSafeName packageName)
    , "packageName" $= mkStr packageName
    , "version"     $= mkStr version
    , "src"         $= fetchurl source
    ]

nixSafeName :: Text -> Text
nixSafeName
  = replace "@" "_at_"
  . replace "/" "_slash_"

fetchurl :: Source -> NExpr
fetchurl (Source url hash) =
  case hash of
    SHA1   h -> "fetchurl" @@ mkNonRecSet [ "url" $= mkStr url, "sha1"   $= mkStr h ]
    SHA512 h -> "fetchurl" @@ mkNonRecSet [ "url" $= mkStr url, "sha512" $= mkStr h ]

mkArgs :: PackageDefinition -> NExpr
mkArgs PackageDefinition { .. } =
  mkNonRecSet
    [ "name"            $= mkStr (nixSafeName packageName)
    , "packageName"     $= mkStr packageName
    , "src"             $= mkPath False src
    , lookupDependencies dependencies
    , "buildInputs"     $= "globalBuildInputs"
    , "meta"            $= emptySet
    , "production"      $= mkBool True
    , "bypassCache"     $= mkBool True
    , "reconstructLock" $= mkBool False
    ]

lookupDependencies :: [Dependency] -> Binding NExpr
lookupDependencies deps = "dependencies" $= mkList (lookupDependency <$> deps)

lookupDependency :: Dependency -> NExpr
lookupDependency = \case
  SimpleDependency    qn         -> lookupSource qn
  RecursiveDependency qn subDeps -> lookupSource qn $//
    mkNonRecSet [ lookupDependencies subDeps ]

lookupSource :: QName -> NExpr
lookupSource qn = "sources" @. renderQName qn

renderQName :: QName -> Text
renderQName (QName qn) = quote qn

fromPackageLock :: PackageLock -> NExpr
fromPackageLock = uncurry nodePackages . (definitionFromPackageLock &&& sourcesFromPackageLock)

definitionFromPackageLock :: PackageLock -> PackageDefinition
definitionFromPackageLock PackageLock { name, packages, src } =
  let mkDep (depName, Package{ version }) =
        let deps = selectRecursiveDependencies depName packages
            qn = QName (depName <> "-" <> version)
         in if M.null deps
            then SimpleDependency qn
            else RecursiveDependency qn (mkDep <$> selectRootPackages deps)
   in PackageDefinition
        { packageName  = name
        , src
        , dependencies = mkDep <$> selectRootPackages packages
        }

sourcesFromPackageLock :: PackageLock -> [PackageSource]
sourcesFromPackageLock PackageLock { packages } =
  let sources = sortOn deriveKey $ M.elems $ M.mapMaybeWithKey buildSource packages
      deriveKey PackageSource { packageName, version } = packageName <> "-" <> version
      buildSource pk Package{..} = do
        hash <- hashFromIntegrity integrity
        pure $
          PackageSource { packageName = packageNameFromPackageKey pk
                        , version
                        , source = Source { url = resolved, hash }
                        }
   in sources

nodePackages :: PackageDefinition -> [PackageSource] -> NExpr
nodePackages packageDefinition sources =
  let mkRequired n   = (n, Nothing)
      mkOptional n v = (n, Just v)
      topLevelParams =
        mkParamSet
          [ mkRequired "nodeEnv"
          , mkRequired "fetchurl"
          , mkRequired "fetchgit"
          , mkRequired "nix-gitignore"
          , mkRequired "stdenv"
          , mkRequired "lib"
          , mkOptional "globalBuildInputs" emptyList
          ]
      bindings =
        [ "sources" $= mkSources sources
        , "args"    $= mkArgs packageDefinition
        ]
      nodeDependenciesArgs = "lib" @. "overrideExisting" @@ "args" @@ mkNonRecSet
        [ "src" $= "stdenv" @. "mkDerivation" @@ mkNonRecSet
            [ "name" $= "args" @. "name" $+ mkStr "-package-json"
            , "src" $= "nix-gitignore" @. "gitignoreSourcePure" @@ mkList
                [ mkStr "*"
                , mkStr "!package.json"
                , mkStr "!package-lock.json"
                ] @@ "args" @. "src"
            , "dontBuild" $= mkBool True
            , "installPhase" $= mkStr "mkdir -p $out; cp -r ./* $out;"
            ]
        ]
      body =
        mkNonRecSet
          [ inherit [ "sources", "args" ]
          , "tarball"          $= "nodeEnv" @. "buildNodeSourceDist"   @@ "args"
          , "package"          $= "nodeEnv" @. "buildNodePackage"      @@ "args"
          , "shell"            $= "nodeEnv" @. "buildNodeShell"        @@ "args"
          , "nodeDependencies" $= "nodeEnv" @. "buildNodeDependencies" @@ nodeDependenciesArgs
          ]
   in mkFunction topLevelParams (mkLets bindings body)


src/PackageLock.hs
{-# LANGUAGE DeriveAnyClass             #-}
{-# LANGUAGE DeriveGeneric              #-}
{-# LANGUAGE DerivingStrategies         #-}
{-# LANGUAGE FlexibleContexts           #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE LambdaCase                 #-}
{-# LANGUAGE OverloadedStrings          #-}
{-# LANGUAGE RecordWildCards            #-}
{-# LANGUAGE TupleSections              #-}

module PackageLock
  ( loadPackageLock
  , PackageLock(..)
  , Packages
  , PackageKey
  , packageNameFromPackageKey
  , Package(..)
  , parseKey
  , selectRootPackages
  , selectRecursiveDependencies
  ) where

import           Data.Aeson
import           Data.Aeson.Types   (parseMaybe)
import qualified Data.List.NonEmpty as NE
import qualified Data.Map.Strict    as M
import           System.FilePath    (takeDirectory)
import           Text.Parsec        (ParsecT, anyChar, choice, eof, many1,
                                     manyTill, runParser, string, try)

data PackageLock =
  PackageLock
    { name            :: Text
    , lockfileVersion :: Int
    , packages        :: Packages
    , src             :: FilePath
    } deriving stock (Show)

loadPackageLock :: FilePath -> IO (Either String PackageLock)
loadPackageLock file =
  let realise (UnrealisedPackageLock unrealised) = unrealised (takeDirectory file)
   in fmap realise <$> eitherDecodeFileStrict file

newtype UnrealisedPackageLock = UnrealisedPackageLock (FilePath -> PackageLock)

instance FromJSON UnrealisedPackageLock where
  parseJSON = withObject "PackageLock" $ \o -> do
    lockfileVersion <- o .: "lockfileVersion"
    if lockfileVersion == 1
    then fail "lockfileVersion 1 not supported"
    else do
      name     <- o .: "name"
      packages <- interpretKeys <$> o .: "packages"
      pure $ UnrealisedPackageLock (\src -> PackageLock{..})

newtype PackageKey = PackageKey (NonEmpty Text) deriving newtype (Show, Eq, Ord, Hashable, FromJSON)

type Packages = Map PackageKey Package

newtype RawPackage = RawPackage { unRawPackage :: Maybe Package }

instance FromJSON RawPackage where
  parseJSON = pure . RawPackage . parseMaybe parseJSON

interpretKeys :: Map Text RawPackage -> Map PackageKey Package
interpretKeys =
  let decodeKey k package =
        maybe id (uncurry M.insert) ((,) <$> parseKey k <*> unRawPackage package)
   in M.foldrWithKey decodeKey mempty

packageNameFromPackageKey :: PackageKey -> Text
packageNameFromPackageKey (PackageKey keys) = last keys

selectRootPackages :: Packages -> [(Text, Package)]
selectRootPackages =
  let pickRoot (PackageKey k) v = (, v) <$> trySingleton k
      trySingleton = \case
        k :| [] -> Just k
        _       -> Nothing
   in M.elems . M.mapMaybeWithKey pickRoot

selectRecursiveDependencies :: Text -> Packages -> Packages
selectRecursiveDependencies packageName =
  let insertSubDependency (PackageKey k) package packages =
        if [packageName] `NE.isPrefixOf` k && length k > 1
        then foldMap (\k' -> M.insert (PackageKey k') package packages) (nonEmpty (tail k))
        else packages
   in M.foldrWithKey insertSubDependency mempty

parseKey :: Text -> Maybe PackageKey
parseKey =
  let runParsec p = either (const Nothing) pure . runParser p () ""
   in (fmap PackageKey . nonEmpty) <=< runParsec packageKeyParser

packageKeyParser :: ParsecT Text () Identity [Text]
packageKeyParser =
  let p = choice
        [ try $ (:) <$> (fromString <$> manyTill anyChar (try (string "/node_modules/"))) <*> p
        , pure <$> try (fromString <$> many1 anyChar <* eof)
        ]
   in string "node_modules/" *> p

data Package =
  Package
    { version   :: Text
    , resolved  :: Text
    , integrity :: Text
    } deriving stock    (Show, Generic)
      deriving anyclass (FromJSON)


src/CLI/Options.hs
module CLI.Options
  ( readOptions
  , Options(..)
  , Paths(..)
  , NodeJSVersion(..)
  ) where

import           Options.Applicative

readOptions :: IO Options
readOptions =
  execParser (info (optionsParser <**> helper <**> simpleVersioner "v0.0.1.0") (fullDesc <> progDesc "node2nix alternative implementation supporting nodeJS 18+"))

optionsParser :: Parser Options
optionsParser =
  Options <$> pathsParser
          <*> noCopyNodeEnvParser
          <*> nodeJSVersionParser

pathsParser :: Parser Paths
pathsParser =
  let pathOption s l v h = strOption (short s <> long l <> help h <> value v)
      -- used to be "default.nix" in original node2nix
      compositionOption = pathOption 'c' "composition" "node-dependencies.nix" "Path to a Nix composition expression allowing someone to deploy the generated Nix packages from the command-line (defaults to: default.nix)"
      outputOption      = pathOption 'o' "output"      "node-packages.nix"     "Path to a Nix expression representing a registry of Node.js packages (defaults to: node-packages.nix)"
      nodeEnvOption     = pathOption 'e' "node-env"    "node-env.nix"          "Path to the Nix expression implementing functions that builds NPM packages (defaults to: node-env.nix)"
      packageLockOption = pathOption 'l' "lock"        "./package-lock.json"   "Path to the package-lock.json file that pinpoints the variants of all dependencies"
   in Paths <$> compositionOption
            <*> outputOption
            <*> nodeEnvOption
            <*> packageLockOption

noCopyNodeEnvParser :: Parser Bool
noCopyNodeEnvParser = not <$> switch (long "no-copy-node-env" <> help "Do not create a copy of the Nix expression that builds NPM packages")

nodeJSVersionParser :: Parser NodeJSVersion
nodeJSVersionParser =
      flag' NodeJS16 (long "nodejs-16" <> help "Provides all settings to generate expression for usage with Node.js 16.x (default is: nodejs-18_x)")
  <|> flag' NodeJS18 (long "nodejs-18" <> help "Provides all settings to generate expression for usage with Node.js 18.x (default is: nodejs-18_x)")
  <|> flag' NodeJS20 (long "nodejs-20" <> help "Provides all settings to generate expression for usage with Node.js 20.x (default is: nodejs-18_x)")
  <|> pure NodeJS18

data Options =
  Options
    { paths         :: Paths
    , noCopyNodeEnv :: Bool
    , nodeJSVersion :: NodeJSVersion
    }

data NodeJSVersion
  = NodeJS16
  | NodeJS18
  | NodeJS20

data Paths =
  Paths
    { compositionFile :: FilePath
    , packagesFile    :: FilePath
    , nodeEnvFile     :: FilePath
    , packageLockFile :: FilePath
    }

simpleVersioner :: String -- ^ Version string to be shown
                -> Parser (a -> a)
simpleVersioner version =
  infoOption version
    (long "version" <> short 'v' <> help "Show version information")
